In [114]:
import pickle
import pandas as pd
import datetime
import os

from dotenv import load_dotenv

from langchain_core.documents import Document
from langchain_core.vectorstores import InMemoryVectorStore
from langchain_ollama import OllamaEmbeddings
from langchain_core.prompts import ChatPromptTemplate
from langchain_google_genai import ChatGoogleGenerativeAI, GoogleGenerativeAIEmbeddings
from langchain.chains import RetrievalQA, RetrievalQAWithSourcesChain
from langchain.agents import AgentExecutor, create_tool_calling_agent, tool

from ragas import evaluate, RunConfig
from ragas import EvaluationDataset
from ragas.dataset_schema import EvaluationResult
from ragas.llms import LangchainLLMWrapper
from ragas.metrics import LLMContextRecall, Faithfulness, FactualCorrectness

In [115]:
load_dotenv()

True

In [116]:
train_df = pd.read_json("train_short.json")

# traindf where wellformedanswers is not empty
train_df.drop(columns=["query_id", "query_type", "wellFormedAnswers"], inplace=True)

# for answer in column answers apply lambda function to get the first element
train_df["answers"] = train_df["answers"].apply(lambda x: x[0] if len(x) > 0 else None)

# split passages into twenty columns (passage_0_text, passage_0_is_selected, passage_1_text, passage_1_is_selected, ...)
for i in range(10):
    passage_text = train_df["passages"].apply(lambda x: x[i]["passage_text"])
    passage_is_selected = train_df["passages"].apply(lambda x: x[i]["is_selected"])
    train_df[f"passage_{i}_selected"] = passage_is_selected
    train_df[f"passage_{i}_text"] = passage_text

# drop columns passage
train_df.drop(columns=["passages"], inplace=True)

# train_df = train_df.iloc[:10]
train_df = train_df[:100]
train_df

,answers,query,passage_0_selected,passage_0_text,passage_1_selected,passage_1_text,passage_2_selected,passage_2_text,passage_3_selected,passage_3_text,...,passage_5_selected,passage_5_text,passage_6_selected,passage_6_text,passage_7_selected,passage_7_text,passage_8_selected,passage_8_text,passage_9_selected,passage_9_text
0,The immediate impact of the success of the man...,)what was the immediate impact of the success ...,1,The presence of communication amid scientific ...,0,The Manhattan Project and its atomic bomb help...,0,Essay on The Manhattan Project - The Manhattan...,0,The Manhattan Project was the name for a proje...,...,0,The Manhattan Project. This once classified ph...,0,Nor will it attempt to substitute for the extr...,0,Manhattan Project. The Manhattan Project was a...,0,"In June 1942, the United States Army Corps of ...",0,One of the main reasons Hanford was selected a...
1,Restorative justice that fosters dialogue betw...,_________ justice is designed to repair the ha...,0,"group discussions, community boards or panels ...",0,punishment designed to repair the damage done ...,0,Tutorial: Introduction to Restorative Justice....,0,"Organize volunteer community panels, boards, o...",...,0,Each of these types of communities—the geograp...,1,The approach is based on a theory of justice t...,0,Inherent in many people’s understanding of the...,0,"Criminal justice, however, is not usually conc...",0,The circle includes a wide range of participan...
2,The customer care number of Amex India is 1800...,amex india customer care number,0,American Express is the world's premier servic...,0,Customer Service Main Page: Get Information: U...,1,Amex Card India Customer Care Phone Number Pho...,0,Air India American Express Gold Card Customer ...,...,0,"Update Address, Payee Name, Bank Account and C...",0,Please mention the first 11 digits of your Ame...,0,"The postal and official address, email address...",0,Corporate Cardmembers can contact the 24 hour ...,0,Through which any customers can easily contact...
3,"Ramen is a quick-cooking noodles, typically se...",definition of ramen,0,"Ramen is of Chinese origin, however it is uncl...",0,"ramen definition, meaning, what is ramen: a Ja...",0,Definition of ramen written for English Langua...,0,Sapporo ramen comes from Japan's northernmost ...,...,0,Wiktionary (0.00 / 0 votes) Rate this definiti...,0,Related dishes. 1 Nagasaki champon. The noodl...,0,[sometimes with sing. v.] Japanese noodles of ...,0,"‘The noodles, most of which we left behind bec...",0,"On the ground floor level, there is a souvenir..."
4,Rachel Carson died because of cancer.,why did rachel carson die,0,A Fish and Wildlife Service National Wildlife ...,0,Though much of their correspondence was destro...,0,The impetus for Silent Spring was a letter wri...,0,How about the environmentalist and writer Rach...,...,0,"Rachel Carson was born on May 27, 1907, on a f...",0,Lived 1907 – 1964. Rachel Carson played a key ...,0,Rachel Carson had a large love for nature and ...,1,Heroes commit feats of courage involving risk ...,0,"To passers-by the mother would say, pointing, ..."
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
95,No Answer Present.,why do my body tries to retain water,0,drink lots of water and when ur body is hydrat...,0,Water retention problems arise when fluid is n...,0,To avoid water retention during a weight loss ...,0,Treatment involves rectifying the underlying c...,...,0,if your pee is yellow when constantly drinking...,0,Why won&#39;t my body retain water? I&#39;m de...,0,"Pregnancy, oral contraceptives such as the pil...",0,1 Upload failed. 2 We are experiencing some p...,0,drink lots of water and when ur body is hydrat...
96,Your eyelashes keeps falling out due to tuggin...,why do my eyelashes keep falling out,0,When I started losing eyelashes… The first tim...,0,"The last phase is the telogen phase, where the...",0,"During the anagen phase, the eyelashes will n

In [117]:
sample_queries = list(train_df['query'])
expected_responses = list(train_df['answers'])

# create a list of every passage, also append the rowid stretched to 4 digits + passage_0_selected number
passages = []
for i, row in train_df.iterrows():
    for j in range(10):
        passage = row[f"passage_{j}_text"]
        passage_is_selected = row[f"passage_{j}_selected"]
        id = f"{int(i):04d}_{passage_is_selected}"
        passages.append((id, passage))

In [118]:
dataset_documents = [
    Document(
        page_content=one_passage,
        metadata={
            "id": one_id,
            "source": one_id
        }
    ) for one_id, one_passage in passages
]

In [119]:
embeddings = OllamaEmbeddings(model="mxbai-embed-large")
# set api keyfor google
os.environ["GOOGLE_API_KEY"] = "AIzaSyDh9mrRRSHaLGo3-8_S3q28vO-F3CeLeuE"
embeddings = GoogleGenerativeAIEmbeddings(model="models/text-embedding-004")
dataset_vector_store = InMemoryVectorStore(embeddings)

In [120]:
dataset_vector_store.add_documents(dataset_documents)

['35ba367f-ee7d-4cb0-81b6-75b83d32df5d',
 '632f60f8-f797-4c49-83cf-a868fb60c53c',
 'c4009b13-aaff-4fdd-ba36-46d0972001e9',
 '19b3e5c3-2940-4a6c-8e53-1f50bf40be89',
 '2b6373e3-a3e7-4b8a-89e0-9d439bef7b7f',
 'a955d2cf-f4c1-4ef8-87b0-5e7930485754',
 '05282e90-647d-470f-8f49-6f1917f321a6',
 '543af683-35ff-45ce-a98c-51a78a1936b8',
 '7ddbfedd-ec68-4e27-9031-826220999da2',
 '4cd4ba11-c0e5-4e2e-b330-f2965b7bda46',
 '18feaab0-1237-4835-9c64-46a20405aa49',
 '522b663c-5d4e-4fbb-b9af-9beb9227ce82',
 '8982cd2f-13a6-4843-aa4a-375b73fb310f',
 '48e5025a-63e3-4203-8015-c4149f9380b8',
 '624166d6-d6df-473e-90e3-b93927d49a11',
 'c5cd9e19-c0f4-49cb-9336-8e2eaf5a2b46',
 '8fb81ab0-cb10-4345-bc27-2125f65b7cb8',
 'ae194267-b00e-4b9d-a2ad-86cf36adf091',
 '0ff27045-3347-4716-82ba-c134ce945313',
 '6a7cefe2-9654-424f-8228-874acc1e3bad',
 'fae74d32-6bc7-4265-afaa-f11c1036ed37',
 '10c50ad5-8264-4e29-a3a7-84dfc8db8f35',
 '8239efdc-a95a-4d4b-b0b2-f4274af1acfa',
 '4c0e2f5e-481f-46bb-8877-e404913d47f7',
 '1ae5c7a2-cf83-

In [121]:
GOOGLE_KEYS = []

for i in range(10):
    possible_name = f"GOOGLE_API_KEY_{i + 1}"
    if os.getenv(possible_name):
        GOOGLE_KEYS.append(os.getenv(possible_name))

key_index = 0
def get_next_google_key():
    global key_index
    key = GOOGLE_KEYS[key_index]
    key_index = (key_index + 1) % len(GOOGLE_KEYS)
    return key

In [122]:
system_prompt = """
You are a helpful reasearch assistant. Your task is to help the user with their research and do the requested tasks no matter how difficult they are.
You should always try to use the rag tools to check if the answer can be found in the database of text passages. Try very hard to find the answer in the database.
You are given access to a database of text passages and a RAG system that can help you access the information about the selected passages.
When you are asked a question, you should first check if the answer is in the database.
If the question is not related to the passages, inform the user that the question is not related to the passages (it will happen rarely) and you cannot help them with that (BUT ONLY AFTER YOU HAVE CHECKED IT USING THE TOOLS!).
If you are unsure whether it is related to the passages, you should assume that it IS. You MUST always assume at first that the question is related to the passages and check it using the tools.
You should use tools rather too much than too little. You should try your best to answer the question using the tools.
You can use the tools as many times as you want. prioritirize using elastic_search_tool over rag_tool, but if you cannot find the answer using elastic_search_tool, you should use rag_tool.
REMEMBER: the passages are about a lot of different topics, so you should not assume that the question is not related to the passages just because it is about a specific topic.
"""

prompt = ChatPromptTemplate.from_messages(
    [
        ("system", system_prompt),
        ("human", "{input}"),
        ("placeholder", "{agent_scratchpad}"),
    ]
)

In [123]:
def answer_question(question: str) -> tuple[str, list[str]]:
    os.environ["GOOGLE_API_KEY"] = get_next_google_key()

    llm = ChatGoogleGenerativeAI(
        model="gemini-2.5-flash-preview-04-17",
        temperature=0,
        max_tokens=None,
        timeout=None,
        max_retries=1,
    )

    rag_chain = RetrievalQAWithSourcesChain.from_chain_type(
        llm=llm,
        chain_type="stuff",
        retriever=dataset_vector_store.as_retriever(search_kwargs={"k": 5}),
    )

    retrieved_context = []

    @tool
    def rag_tool(query: str) -> str:
        """Use this tool to get information from the RAG model if needed.
        This tool allows you to ask questions about the papers selected by the user.
        This tool should return the answer with all the necessary citations.
        Sometimes the tool will not be able to answer the question, in that case, you should try to formulate the question in a different way.
        Most often than not, the problem with this tool's answer will be in your question formulation, so you should try to rephrase it them
        as this tool will try to answer your question very literally.
        """
        output = rag_chain.invoke(query)
        retrieved_context.append(output['sources'])
        return output

    @tool
    def elastic_search_tool(query: str) -> str:
        """Use this tool to get information from the Elasticsearch index if needed.
        This tool allows you to ask questions about the papers selected by the user.
        This tool should return the answer together with the metadata of the paper, by default returning the first 3 results.
        """
        results = dataset_vector_store.similarity_search(query, k=5)

        response = ""
        for i, result in enumerate(results):
            response += f"Result {i+1}:\n"
            response += f"ID: {result.metadata['id']}\n" if "id" in result.metadata else "N/A\n"
            # response += f"Title: {result.metadata['title']}\n" if "title" in result.metadata else "N/A"
            # response += f"Authors: {result.metadata['authors']}\n" if "authors" in result.metadata else "N/A"
            # response += f"DOI: {result.metadata['doi']}\n" if "doi" in result.metadata else "N/A"
            # response += f"Page: {result.metadata['page']}\n\n" if "page" in result.metadata else "N/A\n\n"
            response += f"Content:\n{result.page_content}\n\n\n\n"

        retrieved_context.append(response)
        return response

    tools = [rag_tool, elastic_search_tool]

    agent = create_tool_calling_agent(llm, tools, prompt)
    agent_executor = AgentExecutor(
        agent=agent, tools=tools, verbose=True, max_iterations=15
    )
    output = agent_executor.invoke({"input": question})[
        "output"
    ]

    print(f"Question: {question}")
    print(f"Answer: {output}")

    return output, retrieved_context

In [124]:
def build_evaluation_dataset(queries, references, retriever, formatter):
    data = []
    for query, reference in zip(queries, references):
        response, relevant_docs = answer_question(query)
        data.append({
            "user_input": query,
            "retrieved_contexts": relevant_docs, # [f"Document id {doc.metadata["id"]}:\n{doc.page_content}" for doc in relevant_docs],
            "response": response,
            "reference": reference,
        })
    return EvaluationDataset.from_list(data)


def format_docs(relevant_docs):
    return [f"Document:\n{doc.page_content}\n\n" for doc in relevant_docs]

In [125]:
evaluation_dataset = build_evaluation_dataset(
    list(train_df['query']),
    list(train_df['answers']),
    dataset_vector_store,
    format_docs
)



> Entering new AgentExecutor chain...

Invoking: `elastic_search_tool` with `{'query': 'immediate impact of the success of the Manhattan Project'}`


Result 1:
ID: 0000_0
Content:
The Manhattan Project and its atomic bomb helped bring an end to World War II. Its legacy of peaceful uses of atomic energy continues to have an impact on history and science.



Result 2:
ID: 0000_1
Content:
The presence of communication amid scientific minds was equally important to the success of the Manhattan Project as scientific intellect was. The only cloud hanging over the impressive achievement of the atomic researchers and engineers is what their success truly meant; hundreds of thousands of innocent lives obliterated.



Result 3:
ID: 0000_0
Content:
Essay on The Manhattan Project - The Manhattan Project The Manhattan Project was to see if making an atomic bomb possible. The success of this project would forever change the world forever making it known that something this powerful can be manmade.

In [126]:
datetime_str = datetime.datetime.now().strftime("%Y_%m_%d_%H_%M_%S")

with open(f"evaluation_dataset_{datetime_str}.pickle", "wb") as f:
    pickle.dump(evaluation_dataset, f)

In [127]:
# load from pickle
with open("evaluation_dataset_2025_06_08_20_58_54.pickle", "rb") as f:
    evaluation_dataset = pickle.load(f)

In [128]:
llm = ChatGoogleGenerativeAI(
    model="gemini-2.5-flash-preview-04-17",
    temperature=0,
    max_tokens=None,
    timeout=None,
    max_retries=1,
)

In [129]:
run_config = RunConfig(timeout=60*10, max_workers=1)
# evaluator_llm = LangchainLLMWrapper(llm)

In [130]:
from typing import Optional, Any
from langchain_core.callbacks.manager import CallbackManagerForLLMRun
from langchain_core.language_models.llms import LLM


class TotallyFreeLLM(LLM):
    models: list[ChatGoogleGenerativeAI] = []
    current_model_index: int = 0

    def assign_models(self, keys: list[str], temp: float = 0.3) -> None:
        self.models = [
            ChatGoogleGenerativeAI(temperature=temp, model="gemini-2.5-flash-preview-04-17", api_key=key)
            for key in keys
        ]
        self.current_model_index = 0

    def _call(
        self,
        prompt: str,
        stop: Optional[list[str]] = None,
        run_manager: Optional[CallbackManagerForLLMRun] = None,
        **kwargs: Any,
    ) -> str:
        for _ in range(len(self.models)):
            try:
                response = self.models[self.current_model_index].invoke(prompt)
                self.current_model_index = (self.current_model_index + 1) % len(self.models)
                return str(response.content)
            except Exception as _:
                self.current_model_index = (self.current_model_index + 1) % len(self.models)
        raise ValueError("All models failed to respond.")

    @property
    def _llm_type(self) -> str:
        """Get the type of language model used by this chat model. Used for logging purposes only."""
        return "FreeLLM"

In [131]:
free_llm = TotallyFreeLLM()
free_llm.assign_models(GOOGLE_KEYS)

In [132]:
len(evaluation_dataset)

100

In [133]:
result_correctness: EvaluationResult = evaluate(
    dataset=evaluation_dataset,
    metrics=[FactualCorrectness(), Faithfulness(), LLMContextRecall()],
    llm=free_llm,
    run_config=run_config,
)

result_correctness

Evaluating:   0%|          | 0/300 [00:00<?, ?it/s]

The LLM did not return a valid classification.
The LLM did not return a valid classification.
The LLM did not return a valid classification.


{'factual_correctness(mode=f1)': 0.2215, 'faithfulness': 0.8789, 'context_recall': 0.6134}

200 questions

2000 passages

3 retrieved passages per question

{'factual_correctness(mode=f1)': 0.2972, 'faithfulness': 0.8808, 'context_recall': 0.6350}

100 questions

1000 passages

5 retrieved passages per question

{'factual_correctness(mode=f1)': 0.2215, 'faithfulness': 0.8789, 'context_recall': 0.6134}



In [150]:
import re
i = 0
correct_counter = 0
for i in range(len(evaluation_dataset)):
    if len(evaluation_dataset[i].retrieved_contexts) == 2:
        for id in evaluation_dataset[i].retrieved_contexts[1].split(" "):
            if int(id[:4]) == i and id[5] == "1":
                correct_counter += 1
    elif len(evaluation_dataset[i].retrieved_contexts) == 1:
        pattern = r"(\d{4})_(\d)"
        matches = re.findall(pattern, evaluation_dataset[i].retrieved_contexts[0])
        for match in matches:
            if int(match[0]) == i and match[1] == "1":
                correct_counter += 1
        if len(matches) == 0:
            x = 5/0
correct_counter

49